xrootd/server/server.go

```diff
@@ -0,0 +1,317 @@
+				resp      xrdproto.Marshaler
+				status    xrdproto.ResponseStatus
+			)
+
+			rBuffer := xrdenc.NewRBuffer(reqData)
+			if err := reqHeader.UnmarshalXrd(rBuffer); err != nil {
+				resp, status = newUnmarshalingErrorResponse(err)
+			} else {
+				resp, status = s.handleRequest(sessionID, reqHeader.RequestID, rBuffer)
+			}
+
+			if err := WriteResponse(conn, reqHeader.StreamID, status, resp); err != nil {
+				s.closedMu.RLock()
+				defer s.closedMu.RUnlock()
+				// TODO: wait for active requests to be processed while closing.
+				if !s.closed {
+					s.errorHandler(errors.WithStack(err))
+				}
+				// Abort the connection if an error occurred during
+				// the writing phase because we can't recover from it.
+				return
+			}
+		}()
+	}
+}
+
+func (s *Server) handleHandshake(conn net.Conn) error {
+	data := make([]byte, handshake.RequestLength)
+	if _, err := io.ReadFull(conn, data); err != nil {
+		return err
+	}
+
+	var req handshake.Request
+	rBuffer := xrdenc.NewRBuffer(data)
+	err := req.UnmarshalXrd(rBuffer)
+	if err != nil {
+		return err
+	}
+
+	correctHandshake := handshake.NewRequest()
+	if !reflect.DeepEqual(req, correctHandshake) {
+		return errors.Errorf("xrootd: connection %v: wrong handshake\ngot = %v\nwant = %v", conn.RemoteAddr(), req, correctHandshake)
+	}
+
+	resp, status := s.handler.Handshake()
+	return WriteResponse(conn, xrdproto.StreamID{0, 0}, status, resp)
+}
+
+func newUnmarshalingErrorResponse(err error) (xrdproto.Marshaler, xrdproto.ResponseStatus) {
+	response := xrdproto.ServerError{
+		Code:    xrdproto.InvalidRequestCode,
+		Message: fmt.Sprintf("An error occurred during parsing the request: %v", err),
+	}
+	return response, xrdproto.Error
+}
+
+func (s *Server) handleRequest(sessionID [16]byte, requestID uint16, rBuffer *xrdenc.RBuffer) (xrdproto.Marshaler, xrdproto.ResponseStatus) {
+	switch requestID {
+	case login.RequestID:
+		var request login.Request
+		err := request.UnmarshalXrd(rBuffer)
+		if err != nil {
+			return newUnmarshalingErrorResponse(err)
+		}
+		return s.handler.Login(sessionID, &request)
+	case protocol.RequestID:
+		var request protocol.Request
+		err := request.UnmarshalXrd(rBuffer)
+		if err != nil {
+			return newUnmarshalingErrorResponse(err)
+		}
+		return s.handler.Protocol(sessionID, &request)
+	case dirlist.RequestID:
+		var request dirlist.Request
+		err := request.UnmarshalXrd(rBuffer)
+		if err != nil {
+			return newUnmarshalingErrorResponse(err)
+		}
+		return s.handler.Dirlist(sessionID, &request)
+	default:
+		response := xrdproto.ServerError{
+			Code:    xrdproto.InvalidRequestCode,
+			Message: fmt.Sprintf("Unknown request id: %d", requestID),
+		}
+		return response, xrdproto.Error
+	}
+}
+
+// ReadRequest reads a XRootD request from conn.
+// ReadRequest returns entire payload of the request including header.
+// ReadRequest requires serialization since multiple ReadFull calls are made.
+func ReadRequest(conn net.Conn) ([]byte, error) {
```

Comment:
it doesn't seem like we really need this to be a `net.Conn`.
couldn't we get away with just: `r io.Reader` ?